src/dtx/dtx_common.c

```diff
@@ -737,26 +712,28 @@ dtx_end(struct dtx_handle *dth, struct ds_cont_hdl *cont_hdl,
 	int rc = 0;
 
 	if (dth == NULL || daos_is_zero_dti(&dth->dth_xid))
-		D_GOTO(out, rc);
+		goto out;
 
-	if (result < 0 && dth->dth_dti_cos_count > 0) {
-		/* XXX: For non-leader replica, even if we fail to
-		 *	make related modification for some reason,
-		 *	we still need to commit the DTXs for CoS.
-		 *	Because other replica may have already
-		 *	committed them. For leader case, it is
-		 *	not important even if we miss to commit
-		 *	the CoS DTXs, because they are still in
-		 *	CoS cache, and can be committed next time.
-		 */
-		rc = vos_dtx_commit(cont->sc_hdl, dth->dth_dti_cos,
-				    dth->dth_dti_cos_count);
-		if (rc != 0)
-			D_ERROR(DF_UUID": Fail to DTX CoS commit: %d\n",
-				DP_UUID(cont->sc_uuid), rc);
-	}
+	if (result < 0) {
+		if (dth->dth_dti_cos_count > 0) {
+			/* XXX: For non-leader replica, even if we fail to
+			 *	make related modification for some reason,
+			 *	we still need to commit the DTXs for CoS.
+			 *	Because other replica may have already
+			 *	committed them. For leader case, it is
+			 *	not important even if we fail to commit
+			 *	the CoS DTXs, because they are still in
+			 *	CoS cache, and can be committed next time.
+			 */
+			rc = vos_dtx_commit(cont->sc_hdl, dth->dth_dti_cos,
+					    dth->dth_dti_cos_count);
```

Comment:
so we only do commit-on-share for failed case? (if (result < 0))

Reply:
No, for succeed case, the shared DTXs have already been committed together with the normal update/punch via the shared PMDK transaction.